src/hb-blob.cc

```diff
@@ -545,53 +548,115 @@ hb_blob_create_from_file (const char *file_name)
 
-  // If the file size is 0 and is a regular file, give up
   // See https://github.com/GNOME/glib/blob/f9faac7/glib/gmappedfile.c#L139-L142
-  if (unlikely (st.st_size == 0 && S_ISREG (st.st_mode))) goto fail;
+  if (unlikely (st.st_size == 0 && S_ISREG (st.st_mode)))
+  {
+    DEBUG_MSG (BLOB, nullptr, "Failed as the file size is 0 and is a regular file");
+    close (fd);
+    free (file);
+    return hb_blob_get_empty ();
+  }
 
   file->length = (unsigned long) st.st_size;
-
-#ifdef HAVE_MMAP
   file->contents = (char *) mmap (nullptr, file->length,
 				  writable ? PROT_READ|PROT_WRITE : PROT_READ,
 				  MAP_PRIVATE, fd, 0);
-  if (unlikely (file->contents == MAP_FAILED)) goto fail;
+
+  if (unlikely (file->contents == MAP_FAILED))
+  {
+    DEBUG_MSG (BLOB, nullptr, "Failed to mmap the file");
+    close (fd);
+    free (file);
+    return hb_blob_get_empty ();
+  }
+  close (fd);
+
 #elif defined(_WIN32) || defined(__CYGWIN__)
-  file->mapping = CreateFileMapping ((HANDLE) _get_osfhandle (fd), nullptr,
+  HANDLE fd = CreateFile (file_name, GENERIC_READ, FILE_SHARE_READ, nullptr,
+			     OPEN_EXISTING,
+			     FILE_ATTRIBUTE_NORMAL | FILE_FLAG_OVERLAPPED, nullptr);
+
+  if (unlikely (fd == INVALID_HANDLE_VALUE))
+  {
+    DEBUG_MSG (BLOB, nullptr, "Failed to open the file (win32)");
+    free (file);
+    return hb_blob_get_empty ();
+  }
+
+  file->length = (unsigned long) GetFileSize (fd, nullptr);
+  file->mapping = CreateFileMapping (fd, nullptr,
 				     writable ? PAGE_WRITECOPY : PAGE_READONLY,
 				     0, 0, nullptr);
-  if (unlikely (file->mapping == nullptr)) goto fail;
+  if (unlikely (file->mapping == nullptr))
+  {
+    DEBUG_MSG (BLOB, nullptr, "Failed to create the file mapping");
+    CloseHandle (fd);
+    free (file);
+    return hb_blob_get_empty ();
+  }
 
   file->contents = (char *) MapViewOfFile (file->mapping,
 					   writable ? FILE_MAP_COPY : FILE_MAP_READ,
 					   0, 0, 0);
   if (unlikely (file->contents == nullptr))
   {
-    CloseHandle (file->mapping);
-    goto fail;
+    DEBUG_MSG (BLOB, nullptr, "Failed to create map view of the file");
+    CloseHandle (fd);
+    free (file);
+    return hb_blob_get_empty ();
   }
+  CloseHandle (fd);
+
 #else
-  file->contents = (char *) malloc (file->length);
-  if (unlikely (!file->contents)) goto fail;
-  read (fd, file->contents, file->length);
   mm = HB_MEMORY_MODE_WRITABLE;
+
+  FILE *fd = fopen (file_name, "rb");
+  if (unlikely (!fd))
+  {
+    DEBUG_MSG (BLOB, nullptr, "Failed to open the file (fopen)");
+    free (file);
+    return hb_blob_get_empty ();
+  }
+  fseek (fd, 0, SEEK_END);
+  file->length = ftell (fd);
+  rewind (fd);
+  file->contents = (char *) malloc (file->length);
+  if (unlikely (!file->contents))
+  {
+    DEBUG_MSG (BLOB, nullptr, "Failed to allocate enough space to read the file");
+    free (file);
+    fclose (fd);
+    return hb_blob_get_empty ();
+  }
+  assert (file->length == fread (file->contents, 1, file->length, fd));
```

Comment:
Undecided if needs this highly unlikely assertion

Reply:
This is by no means unlikely. Read can be interrupted for variety of reasons.

Reply:
Thanks, so it should be turned into "if" rather instead assertion or is it enough? (just want to know when to use `assert` also anyway)

Reply:
Correct.

`assert` should be used to check program logic invariants. Eg. that ref_count > 0... if it fails, it's a programmer error.

For fread, there's a variety of reasons it might return short. From the man page:
```
       On  success,  fread()  and  fwrite() return the number of items read or
       written.  This number equals the number of bytes transferred only  when
       size  is 1.  If an error occurs, or the end of the file is reached, the
       return value is a short item count (or zero).

       fread() does not distinguish between end-of-file and error, and callers
       must use feof(3) and ferror(3) to determine which occurred.
```

That's just input error. We should not crash on it.